crates/agent/src/controllers/executor.rs

```diff
@@ -0,0 +1,204 @@
+}
+
+impl automations::Outcome for Outcome {
+    async fn apply(self, txn: &mut sqlx::PgConnection) -> anyhow::Result<Action> {
+        let Outcome {
+            live_spec_id,
+            status,
+            next_run,
+            failures,
+            error,
+            live_spec_deleted,
+        } = self;
+
+        if live_spec_deleted && error.is_none() {
+            assert!(
+                next_run.is_none(),
+                "expected next_run to be None because live spec was deleted"
+            );
+            agent_sql::live_specs::hard_delete_live_spec(live_spec_id, txn)
+                .await
+                .context("deleting live_specs row")?;
+            tracing::debug!(%live_spec_id, "completed controller task for deleted live spec");
+            return Ok(Action::Done);
+        }
+
+        agent_sql::controllers::update_status(
+            txn,
+            live_spec_id,
+            CONTROLLER_VERSION,
+            &status,
+            failures,
+            error.as_deref(),
+        )
+        .await
+        .context("updating controller status")?;
+
+        let action = next_run
+            .map(|n| Action::Sleep(n.compute_duration()))
+            .unwrap_or(Action::Suspend);
+        Ok(action)
+    }
+}
+
+impl<C: ControlPlane + Send + Sync + 'static> Executor for LiveSpecControllerExecutor<C> {
+    const TASK_TYPE: TaskType = automations::task_types::LIVE_SPEC_CONTROLLER;
+
+    type Receive = Event;
+    type State = State;
+    type Outcome = Outcome;
+
+    async fn poll<'s>(
+        &'s self,
+        pool: &'s sqlx::PgPool,
+        task_id: models::Id,
+        _parent_id: Option<models::Id>,
+        state: &'s mut Self::State,
+        inbox: &'s mut std::collections::VecDeque<(models::Id, Option<Self::Receive>)>,
+    ) -> anyhow::Result<Self::Outcome> {
+        let controller_state = fetch_controller_state(task_id, pool).await?;
+        // Note that `failures` here only counts the number of _consecutive_
+        // failures, and resets to 0 on any sucessful update.
+        let (status, failures, error, next_run) =
+            run_controller(state, inbox, &controller_state, &self.control_plane).await;
+
+        Ok(Outcome {
+            live_spec_id: controller_state.live_spec_id,
+            status,
+            failures,
+            error,
+            next_run,
+            live_spec_deleted: controller_state.live_spec.is_none(),
+        })
+    }
+}
+
+#[tracing::instrument(skip_all, fields(
+    live_spec_id = %controller_state.live_spec_id,
+    catalog_name = %controller_state.catalog_name
+))]
+async fn run_controller<C: ControlPlane>(
+    task_state: &mut State,
+    inbox: &mut VecDeque<(Id, Option<Event>)>,
+    controller_state: &ControllerState,
+    control_plane: &C,
+) -> (ControllerStatus, i32, Option<String>, Option<NextRun>) {
+    let mut next_status = controller_state.current_status.clone();
+    tracing::debug!(?inbox, "inbox events");
```

Comment:
confirming intent here:

The sent messages give really nice structured info about why the controller is being run.
Today, they're ignored (except for logging / metrics) because the existing controller factoring doesn't need them.

Tomorrow, we might want to toggle aspects of what the controller's doing based on the received message types. This would involve more refactoring of the controller than we want to do right now (plus we're still supporting legacy agents).

Is that accurate?

Reply:
Yeah, exactly.